server/functions/src/refreshcasestats.ts

```diff
@@ -0,0 +1,371 @@
+    );
+    throw new Error("Unexpected delta in global cases.");
+  }
+}
+
+function caseStatsDocName(stat: who.CaseStats) {
+  // Datastore key name. E.g. '0:' (globalData) or '2:US' (countryData).
+  return `${stat.jurisdictionType}:${stat.jurisdiction}`;
+}
+
+// Save case stats after doing a bit of validation and cleanup.
+async function saveCaseStats(
+  globalData: who.CaseStats,
+  countryData: Map<string, who.CaseStats>
+) {
+  fixPartialLastDayAll(globalData, countryData);
+
+  // Reject unexpected changes in case stats, e.g. too large an increase
+  let loadedData = await db
+    .collection("StoredCaseStats")
+    .doc(caseStatsDocName(globalData))
+    .get();
+  if (loadedData != undefined) {
+    let oldGlobalData = loadedData.data() as who.CaseStats;
+    if (oldGlobalData != null) {
+      totalCasesDeltaCheck(oldGlobalData.cases, globalData.cases);
+    }
+  }
+
+  console.log("Storing to datastore...");
+
+  await db
+    .collection("StoredCaseStats")
+    .doc(caseStatsDocName(globalData))
+    .set(globalData);
+
+  let promises = [];
+  for (let countryStat of countryData.values()) {
+    promises.push(
+      db
+        .collection("StoredCaseStats")
+        .doc(caseStatsDocName(countryStat))
+        .set(countryStat)
+    );
+  }
+  console.log("Awaiting datastore... " + promises.length);
+  await Promise.all(promises);
+  console.log("Datastore set complete.");
+}
+
+async function refreshCaseStatsAsync() {
+  let countryData = await processCaseStats(WHO_CASE_STATS_URL);
+  let globalData = countryData.get("");
+  if (globalData == undefined) {
+    // Should be impossible!
+    return false;
+  }
+  countryData.delete("");
+  try {
+    await saveCaseStats(globalData, countryData);
+  } catch (Error) {
+    console.log("Error saving case stats: " + Error.message);
+    return false;
+  }
+  return true;
+}
+
+const runtimeOpts = {
+  timeoutSeconds: 300,
+};
+
+// Refresh case statistics from data published by WHO on ArcGIS.
+// TODO: Secure so not just anyone can hit it.
+// TODO: ??? Rename to be 'internal' somehow? Use 'background functions'.
+export const refreshCaseStats = functions
+  .region(SERVING_REGION)
+  .runWith(runtimeOpts)
+  .https.onRequest((request, response) => {
+    refreshCaseStatsAsync()
+      .then(function (ok) {
+        if (!ok) {
+          response.status(500).send("Error");
+          return;
+        }
+        response.status(200).send("OK");
+      })
+      .catch(function (error) {
+        console.log(error);
+        response.status(500).send("Error"); // How to get the system to backoff/retry? What text to use?
+        return;
+      });
+  });
+
+// TODO: How do we specify which region this should run in? This data is public,
+// but we presumably want it to run in the same region as our datastore and
+// other code.
+// TODO: Scheduled functions require billing (lol). It's only $0.10/month but
+// but that's absurd for a tiny project...
```

Comment:
It's not even the functions that require billing... It's the scheduler that fires off the Pub/Sub message that _triggers_ the function that needs it... Le sigh. 😕 

FWIW, even on billing-enabled projects the free tier is in effect, so we'll likely still be charged $0.